node/src/components/storage.rs

```diff
@@ -710,6 +705,73 @@ impl Storage {
             .transpose()
     }
 
+    /// Retrieves the highest block from the storage, if one exists.
+    /// May return an LMDB error.
+    fn get_highest_block<Tx: Transaction>(
+        &self,
+        txn: &mut Tx,
+    ) -> Result<Option<Block>, LmdbExtError> {
+        self.block_height_index
+            .keys()
+            .last()
+            .and_then(|&height| self.get_block_by_height(txn, height).transpose())
+            .transpose()
+    }
+
+    /// Returns vector blocks that satisfy the predicate, starting from the latest one and following
+    /// the ancestry chain.
+    fn get_blocks_until<F, Tx: Transaction>(
+        &self,
+        txn: &mut Tx,
+        predicate: F,
+    ) -> Result<Vec<Block>, LmdbExtError>
+    where
+        F: Fn(&Block) -> bool,
+    {
+        let mut next_block = self.get_highest_block(txn)?;
+        let mut blocks = Vec::new();
+        loop {
+            match next_block {
+                None => break,
+                Some(block) if !predicate(&block) => break,
+                Some(block) => {
+                    next_block = match block.parent() {
+                        None => None,
+                        Some(parent_hash) => self.get_single_block(txn, &parent_hash)?,
+                    };
+                    blocks.push(block);
+                }
+            }
+        }
+        Ok(blocks)
+    }
+
+    /// Returns the vector of deploys whose TTL hasn't expired yet.
+    fn get_finalized_deploys(
+        &self,
+        ttl: TimeDiff,
+    ) -> Result<Vec<(DeployHash, DeployHeader)>, LmdbExtError> {
+        let mut txn = self.env.begin_ro_txn()?;
+        // We're interested in deploys whose TTL hasn't expired yet.
+        let ttl_expired = |block: &Block| block.timestamp().elapsed() < ttl;
+        let mut deploys = Vec::new();
+        for block in self.get_blocks_until(&mut txn, ttl_expired)? {
+            for deploy_hash in block.body().deploy_hashes() {
+                let deploy_header = self
+                    .get_deploy_header(&mut txn, &deploy_hash)?
+                    .expect("deploy to exist in storage");
+                deploys.push((*deploy_hash, deploy_header));
+            }
+            for transfer_hash in block.body().transfer_hashes() {
+                let deploy_header = self
+                    .get_deploy_header(&mut txn, &transfer_hash)?
+                    .expect("deploy to exist in storage");
+                deploys.push((*transfer_hash, deploy_header));
```

Comment:
We could also check the TTL here - the deploys can be older than the block, so even if the block's timestamp is less than TTL ago, the deploys can already be expired.

Reply:
Good point! Less data to look through later.

Reply:
Added in [`5035f0d` (#1216)](https://github.com/CasperLabs/casper-node/pull/1216/commits/5035f0d8a3087a306eb5e11a3b354995f7d8707b)